File: packages/typescript/openlibx402-express/src/responses.ts
/**
 * Response Builders for X402
 *
 * Utilities for building properly formatted 402 Payment Required responses.
 */

import { Response } from 'express';
import { PaymentRequest } from '@openlibx402/core';
import { randomBytes } from 'crypto';

export interface Build402ResponseOptions {
  amount: string;
  paymentAddress: string;
  tokenMint: string;
  network: string;
  resource: string;
  description?: string;
  expiresIn?: number; // seconds
}

export function build402Response(
  res: Response,
  options: Build402ResponseOptions
): Response {
  const {
    amount,
    paymentAddress,
    tokenMint,
    network,
    resource,
    description,
    expiresIn = 300,
  } = options;

  const expiresAt = new Date(Date.now() + expiresIn * 1000);

  const paymentRequest = new PaymentRequest({
    max_amount_required: amount,
    asset_type: 'SPL',
    asset_address: tokenMint,
    payment_address: paymentAddress,
    network,
    expires_at: expiresAt,
    nonce: randomBytes(32).toString('base64url'),
    payment_id: randomBytes(16).toString('base64url'),
    resource,
    description,
  });

  return res
    .status(402)
    .set({
      'X-Payment-Required': 'true',
      'X-Payment-Protocol': 'x402',
      'X-Payment-Amount': amount,
      'X-Payment-Asset': tokenMint,
    })
    .json(paymentRequest.toDict());
}

File: packages/python/openlibx402-fastapi/openlibx402_fastapi/responses.py
"""
Response Builders for X402

Utilities for building properly formatted 402 Payment Required responses.
"""
from datetime import datetime, timedelta, timezone
from typing import Optional
import secrets

from fastapi.responses import JSONResponse
from openlibx402_core import PaymentRequest


def build_402_response(
    amount: str,
    payment_address: str,
    token_mint: str,
    network: str,
    resource: str,
    description: Optional[str] = None,
    expires_in: int = 300,
) -> JSONResponse:
    """
    Build a properly formatted 402 Payment Required response

    Args:
        amount: Payment amount required
        payment_address: Recipient wallet address
        token_mint: Token mint address (e.g., USDC)
        network: Network identifier (e.g., "solana-devnet")
        resource: API endpoint being accessed
        description: Human-readable description
        expires_in: Payment validity in seconds

    Returns:
        JSONResponse with 402 status code and payment details
    """
    payment_request = PaymentRequest(
        max_amount_required=amount,
        asset_type="SPL",
        asset_address=token_mint,
        payment_address=payment_address,
        network=network,
        expires_at=datetime.now(timezone.utc) + timedelta(seconds=expires_in),
        nonce=secrets.token_urlsafe(32),
        payment_id=secrets.token_urlsafe(16),
        resource=resource,
        description=description,
    )

    return JSONResponse(
        status_code=402,
        content=payment_request.to_dict(),
        headers={
            "X-Payment-Required": "true",
            "X-Payment-Protocol": "x402",
            "X-Payment-Amount": amount,
            "X-Payment-Asset": token_mint,
        }
    )
